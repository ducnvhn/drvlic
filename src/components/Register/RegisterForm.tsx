import React from 'react'
import {
  makeStyles
} from '@material-ui/styles'
import {
  Grid,
  TextField,
  Paper,
  Typography,
  InputAdornment,
  IconButton,
  Button,
  FormGroup,
  Checkbox,
  FormControlLabel,
  CircularProgress
} from '@material-ui/core'
import {
  Visibility,
  VisibilityOff,
} from '@material-ui/icons'
// import CircularProgress from '@material-ui/icons/'
import {
  useForm,
  Controller,
  SubmitHandler
} from 'react-hook-form'
import {
  gql,
  useMutation
} from '@apollo/client'

export const REGISTER = gql`
  mutation signup($name: String!, $email: String!, $password: String!, $confirmPassword: String!) {
    signup(name: $name, email: $email, password: $password, confirmPassword: $confirmPassword) {
      _id
      role
      email
      name
      profile {
        avatar
      }
      expDate
    }
  }
`
const useStyle = makeStyles({
  regContainer: {
    flexGrow: 1
  },
  formContainer: {
    marginTop: '8em',
    padding: '1em'
  },
  formItem: {
    marginTop: '1em'
  }
})

interface IFormInput {
  email: string
  password: string
  name: string
  confirmPassword: string
  agreement: boolean
}

const RegisterForm = () => {
  const classes = useStyle()
  const { control, handleSubmit, formState: { errors }, watch } = useForm<IFormInput>()
  const [passwordVisible, togglePasswordVisible] = React.useState(false)
  const [registerAccount, { loading }] = useMutation(REGISTER)
   
  const handleClickShowPassword = () => togglePasswordVisible(!passwordVisible)

  const onSubmit:SubmitHandler<IFormInput> = async data => {
    console.log('register btn clicked')
    const { name, email, password, confirmPassword } = data
    if (!name || !email || !password) {
      // show warining here
      return
    }
    if (password !== confirmPassword) {
      // show warning
      return
    }
    if (Object.keys(errors).length > 0) {
      return
    }
    // post data
    await registerAccount({ variables: {
      name,
      email,
      password,
      confirmPassword
    }})
  } 
  // if (errors) {
  //   console.log(errors)
  // }
  return (
    <Grid data-testid="rcontainer" className={classes.regContainer} container alignItems="center" justifyContent="center">
      <Grid xs={12} md={6} lg={4} item>
        <Paper className={classes.formContainer}>
          <Typography style={{ textAlign: 'center' }} variant="h4">Đăng Ký tài khoản</Typography>
          <form onSubmit={handleSubmit(onSubmit)} data-testid="rform">
            <Controller
              rules={{ required: true, minLength: 6 }}
              control={control}
              name="name"
              render={({field}) => (
                <TextField
                  error={'name' in errors}
                  placeholder="Họ và tên người sử dụng"
                  label="Họ và Tên"
                  variant="outlined"
                  className={classes.formItem}
                  fullWidth
                  {...field}
                  helperText={'name' in errors ? '(*) Bắt buộc - Tối thiểu 6 ký tự' : 'Tên có thể viết tiếng Việt có dấu '}
                />
              )}
            />
            <Controller
              control={control}
              rules={{ required: true, minLength: 6 }}
              name="email"
              render={({field}) => (
                <TextField
                error={'email' in errors}
                placeholder="Số điện thoại hoặc email"
                label="Email / SĐT"
                variant="outlined"
                className={classes.formItem}
                fullWidth
                  {...field}
                  helperText={'email' in errors ? '(*) Bắt buộc - phải là email hoặc số điện thoại': 'Nhập email hoặc số điện thoại đề hợp lệ'}
                  />
              )}
            />
            <Controller
              control={control}
              name="password"
              rules={{ required: true, minLength: 8 }}
              render={({ field }) => (
                <TextField
                error={'password' in errors}
                placeholder="Mật khẩu"
                label="Mật khẩu"
                variant="outlined"
                className={classes.formItem}
                fullWidth
                helperText={'password' in errors ? '(*) Bắt buộc - tối thiểu 8 ký tự' : 'Nhập mật khẩu - tối thiểu 8 ký tự'}
                  type={passwordVisible ? 'text' : 'password'}
                  InputProps={{
                    endAdornment:
                    <InputAdornment position="end">
                        <IconButton
                          aria-label="toggle password visibility"
                          onClick={handleClickShowPassword}
                          onMouseDown={handleClickShowPassword}
                          >
                          {passwordVisible ? <Visibility /> : <VisibilityOff />}
                        </IconButton>
                      </InputAdornment>
                    
                  }}
                  {...field}
                />
                )}
                />
            <Controller 
              control={control}
              rules={{ required: true, minLength: 8 }}
              name="confirmPassword"
              render={({ field }) => (
                <TextField
                  placeholder="Nhập lại Mật khẩu"
                  error={'confirmPassword' in errors}
                  label="Nhập lại Mật khẩu"
                  variant="outlined"
                  helperText={'confirmPassword' in errors ? '(*) Bắt buộc - tối thiểu 8 ký tự, trùng khớp bên trên' : 'Nhập mật khẩu - tối thiểu 8 ký tự'}
                  className={classes.formItem}
                  fullWidth
                  type={passwordVisible ? 'text' : 'password'}
                  InputProps={{
                    endAdornment:
                      <InputAdornment position="end">
                        <IconButton
                          aria-label="toggle password visibility"
                          onClick={handleClickShowPassword}
                          onMouseDown={handleClickShowPassword}
                        >
                          {passwordVisible ? <Visibility /> : <VisibilityOff />}
                        </IconButton>
                      </InputAdornment>
                    
                  }}
                  {...field}
                />
              )}
            />
            <br />
            <Controller
              control={control}
              name="agreement"
              render={({ field }) => (
                  <FormGroup row>
                    <FormControlLabel
                      control={<Checkbox {...field} />}
                      label="Tôi đồng ý với các điều khoản sử dụng của e-pharma"
                    />
                  </FormGroup>
                )
              }
            />
            <Button
              data-testid="rbutton"
              className={classes.formItem} variant="contained"
              fullWidth
              color="primary"
              size="large"
              disableElevation
              onClick={handleSubmit(onSubmit)}
              disabled={!watch('agreement') || Object.keys(errors).length > 0}
            >
              {loading ? <CircularProgress data-testid="loadingicon" /> : 'Đăng ký'}
            </Button>
          </form>
          <Typography>
            {`Nếu đã có tài khoản, vui lòng đăng nhập`}
            <a href="/login">{`TẠI ĐÂY`}</a>
          </Typography>
        </Paper>
      </Grid>
    </Grid>
  )
}

export default RegisterForm