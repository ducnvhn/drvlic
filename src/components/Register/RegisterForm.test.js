import React from 'react'
import {
  cleanup,
  render,
  screen,
  // fireEvent
} from '@testing-library/react'
import RegisterForm, { REGISTER } from './RegisterForm'
import { MockedProvider } from '@apollo/client/testing';

const mocks = [
  {
    request: {
      query: REGISTER,
      variables: {
        name: '_',
        email: '_',
        password: '_',
        confirmPassword: '_'
      }
    },
    result: {
      data: {
        signup: {
          _id: '111',
          role:'OWNER',
          email: 'ducnv@gmail.com',
          name: 'ducnv',
          profile: {
            avatar: 'hello there.com'
          },
          expDate:'22/11/2222'
        }
      }
    }
  }
]
afterEach(cleanup)

test('Register form display all components correctly', () => {
  render (
    <MockedProvider mocks={mocks}>
      <RegisterForm />
    </MockedProvider>
  )
  expect(screen.queryByText(/Đăng Ký tài khoản/i)).toBeInTheDocument()
  expect(screen.getByTestId('rform')).toBeInTheDocument()
  expect(screen.getByTestId('rbutton')).toBeInTheDocument()
})

test('show circular progress on click', async () => {
  render (
    <MockedProvider mocks={[]}>
      <RegisterForm />
    </MockedProvider>
  )
  // const btn = screen.getByTestId('rbutton')
  // fireEvent.click(screen.getByTestId('rbutton'))
  // const ldingCir = await screen.findByTestId('loadingicon')
  expect(await screen.findByTestId('rcontainer')).toMatchSnapshot()
})