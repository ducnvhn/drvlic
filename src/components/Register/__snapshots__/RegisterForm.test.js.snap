// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show circular progress on click 1`] = `
<div
  class="MuiGrid-root makeStyles-regContainer-12 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
  data-testid="rcontainer"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 MuiGrid-grid-lg-4"
  >
    <div
      class="MuiPaper-root makeStyles-formContainer-13 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4"
        style="text-align: center;"
      >
        Đăng Ký tài khoản
      </h4>
      <form
        data-testid="rform"
      >
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-formItem-14 MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Họ và Tên
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="name"
              placeholder="Họ và tên người sử dụng"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-15 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-17"
              >
                <span>
                  Họ và Tên
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
          >
            Tên có thể viết tiếng Việt có dấu 
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-formItem-14 MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Email / SĐT
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              name="email"
              placeholder="Số điện thoại hoặc email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-15 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-17"
              >
                <span>
                  Email / SĐT
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
          >
            Nhập email hoặc số điện thoại đề hợp lệ
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-formItem-14 MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Mật khẩu
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              name="password"
              placeholder="Mật khẩu"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                aria-label="toggle password visibility"
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-15 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-17"
              >
                <span>
                  Mật khẩu
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
          >
            Nhập mật khẩu - tối thiểu 8 ký tự
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-formItem-14 MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            Nhập lại Mật khẩu
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              name="confirmPassword"
              placeholder="Nhập lại Mật khẩu"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                aria-label="toggle password visibility"
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-15 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-17"
              >
                <span>
                  Nhập lại Mật khẩu
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
          >
            Nhập mật khẩu - tối thiểu 8 ký tự
          </p>
        </div>
        <br />
        <div
          class="MuiFormGroup-root MuiFormGroup-row"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-19 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-22"
                  data-indeterminate="false"
                  name="agreement"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Tôi đồng ý với các điều khoản sử dụng của e-pharma
            </span>
          </label>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-formItem-14 MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-disableElevation Mui-disabled MuiButton-fullWidth Mui-disabled"
          data-testid="rbutton"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Đăng ký
          </span>
        </button>
      </form>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Nếu đã có tài khoản, vui lòng đăng nhập
        <a
          href="/login"
        >
          TẠI ĐÂY
        </a>
      </p>
    </div>
  </div>
</div>
`;
